### PR TITLE
Take into account explicit hydrogens when creating and representing Hydrogen bonds (regular and weak)

### LINKED ISSUE #1163: Hydrogen bonds perceived between polar hydrogens
Hi team!
I was looking at PDB 1GJ5 and noticed that the hydrogens from a water molecules had Hbonds interactions perceived with other polar hydrogens (from hydroxyl or amine groups).
image

1GJC displays another example of questionable HBonds network:
image

I would expect that no HBonds could be perceived between H atoms in general.
I suppose that HBonds are computed primarily from the heavy atoms and further adjusted if Hydrogen atoms are present. The results look good in some other structures of the same series (1GJD).
Thanks!

### PR DESCRIPTION
**Pull request by papillot**
# Description
Fixes #1163 
This PR refines the way Hydrogen bonds are created and represented when explicit hydrogen atoms atoms are present:
- Donors: the angle between H-Donor-Acceptor must be below the Donor Deviation angle value. This ensures that when H atoms are present on the donor, there is at least one atom positioned in a conic region of angle 90º between the donor and the acceptor
- Acceptors: the H atoms are taken into account to satisfy the geometry of the bond towards the donor. This is the same criteria as for heavy atoms bonded to the acceptor. It ensures that there are no hydrogen atoms on the acceptor that are "in the way" of the hydrogen bond. In other words, the hydrogen bonds is anchored on an unoccupied area of the acceptor.
- Representations: links are made using the closest hydrogen atom of the donor atom (not on the acceptor) when hydrogen atoms are represented. If only polar hydrogens are represented, they are used for regular hydrogen bonds, but weak hydrogens bonds are anchored on the C donor directly.
- A new geometry option has been added "Ignore Hydrogens". This takes into account the hydrogen atoms position for calculating the interactions. It is set to `false` by default. My understanding is that this option is different than the one that controls the display of the hydrogen atoms, as I can think that someone might wish to hide the hydrogen atoms but still benefit from using the explicit hydrogens to display only the most "realistic" hydrogen bonds. And in some cases, if the hydrogens atoms have been places arbitrarily, one might wish to not base interactions on their positions.
Screenshot 2024-06-25 at 11 30 12

## Comparison with other implementations
Using PDB code 1GJ5 and looking at the interactions with ligand [130] (using PDB file as input)
| Viewers   | Total HBonds | Water Hbonds | Sulfur HBond |
|-----------|--------------|----------------|--------------|
| MolStar     | 7                    | 2                        | 1                      |
| ChimeraX | 5                     | 1                        | 0                     |
| PyMol       | 5                     | 2                        | 0                    |
|MolStar 4.3| 13                  | 4                        | 1                     |

The most dubious one is this water hydrogen bond:
image
(note that the covalent bond from the ligand oxygen is not in the PDB file, and not represented in the other viewers)

## Actions

- [x] Added description of changes to the `[Unreleased]` section of `CHANGELOG.md`
- [x] Updated headers of modified files
- [x] Added my name to `package.json`'s `contributors`
- [ ] (Optional but encouraged) Improved documentation in `docs`

### CODE REVIEW

**Comment by arose** [comment]
Nice, will have a look later this week

File: src/mol-model-props/computed/representations/interactions-intra-unit-cylinder.ts [hunk 1]
```diff
@@ -66,9 +67,11 @@ async function createIntraUnitInteractionsCylinderMesh(ctx: VisualContext, unit:
                 let minDistB = minDistA;
                 Vec3.copy(posA, pA);
                 Vec3.copy(posB, pB);
+                const donorType = t === InteractionType.HydrogenBond ? FeatureType.HydrogenDonor : FeatureType.WeakHydrogenDonor;
+                const isHydrogenDonorA = types[offsets[a[edgeIndex]]] === donorType;
 
-                eachIntraBondedAtom(unit, idxA, (_, idx) => {
-                    if (isHydrogen(structure, unit, elements[idx], 'polar')) {
+                if (isHydrogenDonorA) eachIntraBondedAtom(unit, idxA, (_, idx) => {
```
Comment (arose) [inline src/mol-model-props/computed/representations/interactions-intra-unit-cylinder.ts:73]: please add curly brances to the if-else statement

File: CHANGELOG.md [hunk 2]
```diff
@@ -34,6 +34,7 @@ Note that since we don't clearly distinguish between a public and private interf
     - `celShaded` geometry parameter
     - `celSteps` renderer parameter
 - Add the ability to customize the Snapshot Description component via `PluginUISpec.components.viewport.snapshotDescription`
+- Take into account explicit hydrogens when building hydrogen bonds
```
Comment (arose) [inline CHANGELOG.md:37]: building -> computing

File: src/mol-model-props/computed/representations/interactions-intra-unit-cylinder.ts [hunk 3]
```diff
@@ -2,6 +2,7 @@
  * Copyright (c) 2019-2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
```
Comment (arose) [inline src/mol-model-props/computed/representations/interactions-intra-unit-cylinder.ts:2]: please update the date here and in the other files

File: src/mol-model-props/computed/interactions/hydrogen-bonds.ts [hunk 4]
```diff
@@ -230,26 +233,32 @@ type HydrogenBondsOptions = ReturnType<typeof getHydrogenBondsOptions>
...
 
     if (don.idealGeometry[donIndex] === AtomGeometry.Trigonal) {
         const outOfPlane = calcPlaneAngle(structure, don.unit, donIndex, acc.unit, accIndex);
         if (outOfPlane !== undefined && outOfPlane > opts.maxDonOutOfPlaneAngle) return;
     }
 
-    const accAngles = calcAngles(structure, acc.unit, accIndex, don.unit, donIndex);
+    let donnorIndex = donIndex;
+    if (!opts.ignoreHydrogens && donHAngles.length > 0) {
+        donnorIndex = closestHydrogenIndex(structure, don.unit, donIndex, acc.unit, accIndex);
```
Comment (arose) [inline src/mol-model-props/computed/interactions/hydrogen-bonds.ts:253]: donorIndex

**Review by arose — CHANGES_REQUESTED** [review]
This look good, thank you! I would like to do some more tests or see some more examples (if feasible) before merging it.

File: CHANGELOG.md (hunk 2 above)
Comment (papillot) [reply CHANGELOG.md:37]: done

File: src/mol-model-props/computed/representations/interactions-intra-unit-cylinder.ts (hunk 3 above)
Comment (papillot) [reply src/mol-model-props/computed/representations/interactions-intra-unit-cylinder.ts:2]: Done

File: src/mol-model-props/computed/interactions/hydrogen-bonds.ts (hunk 4 above)
Comment (papillot) [reply src/mol-model-props/computed/interactions/hydrogen-bonds.ts:253]: done

File: src/mol-model-props/computed/representations/interactions-intra-unit-cylinder.ts (hunk 1 above)
Comment (papillot) [reply src/mol-model-props/computed/representations/interactions-intra-unit-cylinder.ts:73]: done

File: src/mol-model-props/computed/interactions/hydrogen-bonds.ts (hunk 4 above)
Comment (papillot) [reply src/mol-model-props/computed/interactions/hydrogen-bonds.ts:253]: done

**Comment by papillot** [comment]
> This look good, thank you! I would like to do some more tests or see some more examples (if feasible) before merging it.

Thanks!
Agreed, that would be preferable.

**Comment by papillot** [comment]
Example: 1GAH, around MAN 473
Molstar current version:
Screenshot 2024-07-01 at 10 11 20

Molstar this PR
image

The interactions from the current version encircled in gold, look wrong: between H atoms, or between 2 acceptors.
The one at the bottom of the picture could be questioned: orientation of the lone pair on the oxygen looks correct, but the geometry of the hydrogen on the donor does not support an alignment with the acceptor. Pymol does not support it either.

**Comment by papillot** [comment]
Example: 1st frame of trajectory from GPCR md db: https://www.gpcrmd.org/dynadb/dynamics/id/85/
Focus on ASN64 Chain B

Current Mol* version
Screenshot 2024-07-01 at 10 41 26

This PR:
Screenshot 2024-07-01 at 10 44 38

**Review by arose — APPROVED** [review]
thank you, very nice improvement

good to see a MD sim example

**Pull request merged**